plugins/framework/tests/test_model_patcher.py

```diff
@@ -0,0 +1,136 @@
+# Copyright The IBM Tuning Team
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# SPDX-License-Identifier: Apache-2.0
+# https://spdx.dev/learn/handling-license-info/
+
+# Third Party
+import pytest  # pylint: disable=(import-error
+import torch
+
+# First Party
+from fms_acceleration.model_patcher import (
+    ModelPatcher,
+    ModelPatcherRule,
+    ModelPatcherTrigger,
+    patch_target_module,
+)
+from fms_acceleration.utils.test_utils import DummyModule
+
+DUMMY_RULE_ID = "test_patch"
+DUMMY_HIDDEN_DIM = 32
+
+
+class DummyCrossEntropyLoss(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    def forward(self, inputs, targets):
+        return -1
+
+
+@pytest.fixture()
+def model():
+    return DummyModule(DUMMY_HIDDEN_DIM)
+
+@pytest.fixture()
+def model_inputs(seed: int = 42):
+    torch.manual_seed(seed)
+    return torch.rand(1, DUMMY_HIDDEN_DIM)
+
+def test_rule_registration_and_simple_forward_patching(model_inputs, model): # pylint: disable=redefined-outer-name
+    "Test model patcher replaces the forward function with a dummy forward"
+    # 1. Register rule and specify a trigger on target module for the rule to be applied
+    # 2. Patch model
+    # 3. check target module's forward function and dummy patch produces similar outputs
+    dummy_forward_to_patch = lambda self, X: X * 2 # pylint: disable=unnecessary-lambda-assignment
+    ModelPatcher.rules.pop(DUMMY_RULE_ID, None)
+    rule = ModelPatcherRule(
+        rule_id=DUMMY_RULE_ID,
+        trigger=ModelPatcherTrigger(check=DummyModule),
+        forward=dummy_forward_to_patch,
+    )
+    ModelPatcher.register(rule)
+    assert DUMMY_RULE_ID in ModelPatcher.rules.keys(), "Rule Registration Failed" # pylint: disable=consider-iterating-dictionary
+    ModelPatcher.patch(model)
+    assert torch.allclose(
+        model(model_inputs), model_inputs * 2
+    ), "Failed to patch forward function"
+
+
+# Test patching of model attribute
+def test_patching_downstream_module(model): # pylint: disable=redefined-outer-name
+    "Test patching an imported module indirectly managed by other modules using import_and_reload"
+    # 1. Register rule targeting downstream module and specify target to reload with patch applied
+    # 2. Patch model
+    # 3. check patched module now exist in model
+    ModelPatcher.rules.pop(DUMMY_RULE_ID, None)
```

Comment:
why is this pop needed?

Reply:
if its to changeout the rules, then better to follow the `instantiate framwork` https://github.com/foundation-model-stack/fms-acceleration/blob/main/plugins/framework/src/fms_acceleration/utils/test_utils.py